core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantExecutionStats.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import io.airlift.log.Logger;
+import io.airlift.stats.TimeStat;
+import io.trino.execution.ExecutionFailureInfo;
+import io.trino.execution.TaskInfo;
+import io.trino.execution.TaskState;
+import io.trino.operator.TaskStats;
+import io.trino.spi.ErrorCode;
+import io.trino.spi.ErrorType;
+import io.trino.spi.TrinoException;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import java.util.Optional;
+
+import static io.trino.spi.ErrorType.INTERNAL_ERROR;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.trino.util.Failures.toFailure;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class FaultTolerantExecutionStats
+{
+    private static final Logger log = Logger.get(FaultTolerantExecutionStats.class);
+
+    private final ExecutionStats finishedTasks = new ExecutionStats();
+    private final FailedTasksStats failedTasks = new FailedTasksStats();
+
+    public void update(TaskInfo info)
+    {
+        TaskState state = info.getTaskStatus().getState();
+        switch (state) {
+            case FINISHED:
+                finishedTasks.update(info.getStats());
+                break;
+            case FAILED: {
+                failedTasks.update(info);
+                break;
+            }
+            case CANCELED:
+            case ABORTED:
+                // ignore cancelled and aborted tasks
+                break;
```

Comment:
I know currently dependent tasks are not executing until upstream task comptes. Yet it will change. In such case would we make downstream task as "FAILED" or "ABORTED".
If the latter then we should also compute stats for "ABORTED" tasks. It would be important to understand how much effort we are wasting on those.
Maybe we can just merge FAILED and ABORTED?

Reply:
I'm going to add `private final ExecutionStats abortedTasks = new ExecutionStats();` and store the stats for both, `ABORTED` and `CANCELLED` tasks there. Not sure if it makes sense to track metrics separately.